tardis/energy_input/main_gamma_ray_loop.py

```diff
@@ -4,77 +4,125 @@
 import numpy as np
 import pandas as pd
 
-from tardis.energy_input.energy_source import (
-    get_nuclear_lines_database,
-)
 from tardis.energy_input.gamma_packet_loop import gamma_packet_loop
-from tardis.energy_input.gamma_ray_channel import (
-    calculate_total_decays,
-    create_inventories_dict,
-    create_isotope_dicts,
-)
-
-from tardis.energy_input.gamma_ray_transport import (
-    calculate_total_decays_old,
-    create_isotope_dicts_old,
-    create_inventories_dict_old,
-)
-from tardis.energy_input.gamma_ray_packet_source import RadioactivePacketSource
+from tardis.energy_input.gamma_ray_packet_source import GammaRayPacketSource
 from tardis.energy_input.gamma_ray_transport import (
-    calculate_average_energies,
-    calculate_average_power_per_mass,
     calculate_ejecta_velocity_volume,
-    calculate_energy_per_mass,
-    decay_chain_energies,
-    distribute_packets,
     get_taus,
     iron_group_fraction_per_shell,
 )
 from tardis.energy_input.GXPacket import GXPacket
+from tardis.energy_input.util import make_isotope_string_tardis_like
 
 logger = logging.getLogger(__name__)
 logging.basicConfig(level=logging.INFO)
 
+def get_effective_time_array(time_start, time_end, time_space, time_steps):
+    """
+    Function to get the effective time array for the gamma-ray loop.
+
+    Parameters
+    ----------
+    time_start : float
+        start time in days.
+    time_end : float
+        end time in days.
+    time_space : str
+        linear or log.
+    time_steps : int
+        number of time steps.
+
+    Returns
+    -------
+    times : np.ndarray
+        array of times in secs.
+    effective_time_array : np.ndarray
+        effective time array in secs.
+    """
+
+    assert time_start < time_end, "time_start must be smaller than time_end!"
+    if time_space == "log":
+        times = np.geomspace(time_start, time_end, time_steps + 1)
+        effective_time_array = np.sqrt(times[:-1] * times[1:])
+    else:
+        times = np.linspace(time_start, time_end, time_steps + 1)
+        effective_time_array = 0.5 * (times[:-1] + times[1:])
+
+    return times, effective_time_array
+
+
 
 def run_gamma_ray_loop(
     model,
-    plasma,
+    isotope_decay_df,
+    cumulative_decays_df,
     num_decays,
-    time_start,
-    time_end,
-    time_space,
-    time_steps,
+    times,
+    effective_time_array,
     seed,
     positronium_fraction,
-    atom_data,
     spectrum_bins,
+    time_steps,
```

Comment:
If times are input, we should not allow setting the number of time steps. Instead make it equal to len(times) or similar.

Reply:
Yes. Have considered the suggestion.